Translate useExternalStore reference (#619)
reactjs/ja.react.dev: src/content/reference/react/useSyncExternalStore.md

}
```

これは、ストアにあるデータのスナップショットを返します。引数として 2 つの関数を渡す必要があります：

1. `subscribe` 関数はストアへのサブスクライブを開始します。サブスクライブを解除する関数を返す必要があります。 
2. `getSnapshot` 関数は、ストアからデータのスナップショットを読み取る必要があります。

[さらに例を見る](#usage)

#### 引数 {/*parameters*/}

> * `subscribe`: ストアにサブスクライブを開始し、また callback 引数を受け取る関数。ストアが変更された際に渡された callback を呼び出す必要があります。これにより、コンポーネントが再レンダーされます。subscribe 関数は、サブスクリプションをクリーンアップする関数を返す必要があります。

Reviewer: ```suggestion
* `subscribe`: ストアにサブスクライブを開始し、また callback 引数を受け取る関数。ストアが変更された際に渡された callback を呼び出す必要があります。これにより、コンポーネントが再レンダーされます。`subscribe` 関数は、サブスクリプションをクリーンアップする関数を返す必要があります。
```

> * `getSnapshot`: コンポーネントが必要とするストアにあるデータのスナップショットを返す関数。ストアが変更されていない場合、getSnapshot への再呼び出しは同じ値を返す必要があります。ストアが変更されて返された値が（[`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) で比較して）異なる場合、React はコンポーネントを再レンダーします。

Reviewer: ```suggestion
* `getSnapshot`: コンポーネントが必要とするストアにあるデータのスナップショットを返す関数。ストアが変更されていない場合、`getSnapshot` への再呼び出しは同じ値を返す必要があります。ストアが変更されて返された値が（[`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) で比較して）異なる場合、React はコンポーネントを再レンダーします。
```

```

```js todoStore.js
// これはサードパーティのストアの例です。
// React と統合する必要があるかもしれません。

// アプリが完全に React でビルドされている場合は、 

> // 代わりに React state を使用することをお勧めします。

Author: 8c6c2ef のコミットで対応しました。
確認お願いします。

<Note>

可能であれば、React 組み込みの state 管理機能である [`useState`](/reference/react/useState) および [`useReducer`](/reference/react/useReducer) を代わりに使用することをお勧めします。`useSyncExternalStore` API は、既存の非 React コードと統合する必要がある場合に主に役立ちます。

</Note>

---

### ブラウザ API へのサブスクライブ {/*subscribing-to-a-browser-api*/}

> `useSyncExternalStore` を追加するもう 1 つの理由は、時間とともに変化するブラウザが公開する値にサブスクライブしたい場合です。たとえば、コンポーネントがネットワーク接続がアクティブかどうかを表示したいとします。ブラウザは、この情報を [`navigator.onLine`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine) というプロパティを介して公開します。

Reviewer: ```suggestion
`useSyncExternalStore` を追加するもう 1 つの理由は、時間とともに変化する、ブラウザが公開する値にサブスクライブしたい場合です。たとえば、コンポーネントがネットワーク接続がアクティブかどうかを表示したいとします。ブラウザは、この情報を [`navigator.onLine`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine) というプロパティを介して公開します。
```

}

function getServerSnapshot() {
  return true; // サーバーが生成した HTML には常に "Online "を表示します
}

function subscribe(callback) {
  // ...
}
```

`getServerSnapshot` 関数は `getSnapshot` と似ていますが、以下の 2 つの状況でのみ実行されます：

- HTML を生成する際にサーバ上で実行されます。
- React がサーバ HTML をインタラクティブにするとき、つまり [hydration](/reference/react-dom/client/hydrateRoot) 中にクライアントで実行されます。

これにより、アプリがインタラクティブになる前に使用されう初期のスナップショット値を指定できます。サーバレンダリングに意味のある初期値がない場合は、この引数を省略して、[クライアントで強制的にレンダリングする](/reference/react/Suspense#providing-a-fallback-for-server-errors-and-server-only-content) ようにします。

<Note>

> `getServerSnapshot` が最初のクライアントレンダリングで、サーバで返したのと同じ正確なデータを返すことを確認してください。例えば、`getServerSnapshot` がサーバ上で事前に準備されたストアコンテンツを返した場合、このコンテンツをクライアントに転送する必要があります。これを行う 1 つの方法は、サーバレンダリング中に `window.MY_STORE_DATA` のようなグローバルを設定する `<script>` タグを発行し、`getServerSnapshot` でクライアントのそのグローバルから読み込むことです。外部ストアにその方法が記載されているはずです。

Reviewer: ```suggestion
初回のクライアントレンダリングでは、`getServerSnapshot` はサーバで返したものと必ず正確に同一のデータを返すようにしてください。例えば、`getServerSnapshot` がサーバ上で事前に準備されたストアコンテンツを返した場合、このコンテンツをクライアントに転送する必要があります。これを行う 1 つの方法は、サーバレンダリング中に `window.MY_STORE_DATA` のようなグローバル変数を設定する `<script>` タグを発行しておき、クライアントの `getServerSnapshot` でそのグローバル変数から読み込むことです。あなたが使う外部ストアにその方法が記載されているはずです。
```

</Note>

---

> ## トラブルシューティング {/troubleshooting/} {/*トラブルシューティング-troubleshooting*/}

Reviewer: ```suggestion
## トラブルシューティング {/*troubleshooting*/}
```

> ### エラーが出ています: "getSnapshot の結果はキャッシュされるべきです" {/im-getting-an-error-the-result-of-getsnapshot-should-be-cached/} {/*エラーが出ています-getsnapshot-の結果はキャッシュされるべきです-im-getting-an-error-the-result-of-getsnapshot-should-be-cached*/}

Reviewer: ```suggestion
### "The result of `getSnapshot` should be cached" というエラーが出る {/*im-getting-an-error-the-result-of-getsnapshot-should-be-cached*/}
```

> このエラーは、`getSnapshot` 関数が呼ばれるたびに新しいオブジェクトを返すことを意味します。例えば：

Reviewer: ```suggestion
このエラーは、`getSnapshot` 関数が呼ばれるたびに新しいオブジェクトを返していることを意味します。例えば：
```

```js {2-5}
function getSnapshot() {

>   // 🔴 getSnapshot から常に異なるオブジェクトを返しません

Reviewer: ```suggestion
  // 🔴 Do not return always different objects from getSnapshot
```

  return {
    todos: myStore.todos
  };
}
```

> `getSnapshot` の戻り値が前回と異なる場合、React はコンポーネントを再レンダーします。このため、常に異なる値を返すと無限ループに入り、次のようなエラーが発生します。

Reviewer: ```suggestion
`getSnapshot` の返り値が前回と異なる場合、React はコンポーネントを再レンダーします。このため、常に異なる値を返すと無限ループに入り、このエラーが発生します。
```

`getSnapshot` オブジェクトは、実際に何かが変更された場合にのみ、別のオブジェクトを返す必要があります。ストアにイミュータブルなデータが含まれている場合は、そのデータを直接返すことができます：

```js {2-3}
function getSnapshot() {

>   // ✅ イミュータブルなデータを返すことができます

Reviewer: ```suggestion
  // ✅ You can return immutable data
```

  return myStore.todos;
}
```

> ストアデータが変更可能な場合、`getSnapshot` 関数はその不変スナップショットを返す必要があります。つまり、新しいオブジェクトを作成する必要が**あります**が、毎回呼び出す必要はありません。その代わりに、最後に計算されたスナップショットを保存し、ストア内のデータが変更されていない場合は前回と同じスナップショットを返すようにします。ミュータブルなデータが変更されたかどうかをどのように判断するかは、ミュータブルなストアに依存します。

Reviewer: ```suggestion
ストアデータがミュータブルな場合、`getSnapshot` 関数はそのイミュータブルなスナップショットを返す必要があります。つまり、新しいオブジェクトを作成する必要は**あります**が、毎回作成してはいけないということです。その代わりに、最後に計算されたスナップショットを保存しておき、ストア内のデータが変更されていない場合は前回と同じスナップショットを返すようにします。ミュータブルなデータが変更されたかどうかを判断する方法は、ミュータブルなストアによって異なります。
```

---

> ### `subscribe` 関数が毎回の再レンダー後に呼び出されます {/my-subscribe-function-gets-called-after-every-re-render/} {/*subscribe-関数が毎回の再レンダー後に呼び出されます-my-subscribe-function-gets-called-after-every-re-render*/}

Reviewer: ```suggestion
### `subscribe` が毎レンダーごとに呼び出される {/*my-subscribe-function-gets-called-after-every-re-render*/}
```

> この `subscribe` 関数はコンポーネントの**内部**で定義されるため、再レンダーするたびに異なります：

Reviewer: ```suggestion
この `subscribe` 関数はコンポーネントの**内部**で定義されているため、再レンダーするたびに異なった値になります：
```

```js {4-7}
function ChatIndicator() {
  const isOnline = useSyncExternalStore(subscribe, getSnapshot);
  

>   // 🚩 常に異なる関数であるため、React は再レンダーのたびに再サブスクライブします

Reviewer: ```suggestion
  // 🚩 Always a different function, so React will resubscribe on every re-render
```

  function subscribe() {
    // ...
  }

  // ...
}
```

> React は、再レンダーの間に別の `subscribe` 関数を渡すと、ストアに再サブスクライブします。これがパフォーマンスの問題を引き起こし、再サブスクライブを避けたい場合は、`subscribe` 関数を外部に移動してください：

Reviewer: ```suggestion
React は、再レンダー間で異なる `subscribe` 関数を渡すと、ストアに再サブスクライブします。これがパフォーマンスの問題を引き起こし、再サブスクライブを避けたい場合は、`subscribe` 関数を外部に移動してください：
```

```js {6-9}
function ChatIndicator() {
  const isOnline = useSyncExternalStore(subscribe, getSnapshot);
  // ...
}

> // ✅ 常に同じ関数です、なので React は再サブスクライブする必要がありません

Reviewer: ```suggestion
// ✅ Always the same function, so React won't need to resubscribe
```

function subscribe() {
  // ...
}
```

あるいは、`subscribe` を [`useCallback`](/reference/react/useCallback) でラップすることで、引数が変更されたときのみ再サブスクライブすることができます：

```js {4-8}
function ChatIndicator({ userId }) {
  const isOnline = useSyncExternalStore(subscribe, getSnapshot);
  

>   // ✅ userId が変わらない限り同じ関数になります

Reviewer: ```suggestion
  // ✅ Same function as long as userId doesn't change
```
